### PR TITLE
New music

### LINKED ISSUE #436: New music
# **Description**
Change the music system to introduce new music and allow for new music in the future.
- Replace the OG box boy background music with my epic 8-bit music
- Refactor music playing to accept a music type (normal, boss), making it easier to switch between music types mid-game without having to keep track of asset paths.

# **Dependencies**
- Maps team will likely be adding boss music, I will try to design the code so they only have to change one string for the asset path.

# **Milestones**
- [x] Create and add the new default music to the repo (10 Oct)
- [x] Add code for music switching (11 Oct)
- [ ] Manual testing (11 Oct)
- [ ] PR (11 Oct)
- [ ] Documentation, sprint form (11 Oct)

**Completion Deadline:** 12 Oct

# **Extras (optional)**
- Even more music e.g. menu music? Death screen music?

# **Documentation**
In progress
- [Main description of feature](../wiki/Entity-Component-System-(ECS))
- [User documentation](../wiki/Terrain)
- ...

# **Members**
Ian Pinto @IanP96

### PR DESCRIPTION
**Pull request by IanP96**
# Description

- Change the music (OG music file is still there)
- Add music when you die
- To play boss music:
  - from `MainGameArea`, go to `MainGameArea.MusicType`, change the filename of the `BOSS` music, then add a call to `MainGameArea.playMusic(MusicType)`
  - from any class, use `ResourceService.playMusic(String)`

Closes #436 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

- Unit test for music playing
- Start the game. Music will play on loop.
- Exit the game. Music will stop
- Die. Normal music will stop, death music will play.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by Aryan-S-Gupta — COMMENTED** [review]
I have reviewed and tested the changes you made to refactor the music handling in MainGameArea and BossRoom. The following updates were successfully implemented:

- The playMusic method has been fixed to ensure smooth music transitions.
- The changeRooms method now correctly switches music based on the room type.
- The stopCurrentMusic method has been added for better music management.
- The dispose method has been updated to clean up music resources effectively.
- Added Boss room Music (Temporary to test the functioning)

Please review the changes in your branch and check the implementation and other than that looks fine to me ! This will be approved in the order of Pull Request

**Review by Aryan-S-Gupta — APPROVED** [review]

File: source/core/src/main/com/csse3200/game/areas/MainGameArea.java [hunk 1]
```diff
@@ -309,13 +316,45 @@ private enum MusicType {
...
+     * Stops the currently playing music.
+     */
+    private void stopCurrentMusic() {
+        if (normalMusic != null) {
+            normalMusic.stop();
+        }
+        if (bossMusic != null) {
+            bossMusic.stop();
+        }
     }
```
Comment (IanP96) [inline source/core/src/main/com/csse3200/game/areas/MainGameArea.java:358]: ResourceService.stopCurrentMusic should already do this

**Comment by CameronBadman** [comment]
@IanP96 can I get you to move music stuff into the GameArea, not the MainGameArea?

File: source/core/src/main/com/csse3200/game/areas/MainGameArea.java [hunk 2]
```diff
@@ -309,13 +316,45 @@ private enum MusicType {
...
+    public void playMusic(MusicType musicType) {
+        if (currentMusicType == musicType) {
+            return; // Music is already playing, no need to change
+        }
+
+        stopCurrentMusic();
+
+        Music newMusic = ServiceLocator.getResourceService().getAsset(musicType.path, Music.class);
+        newMusic.setLooping(true);
+        newMusic.play();
```
Comment (IanP96) [inline source/core/src/main/com/csse3200/game/areas/MainGameArea.java:337]: ResourceService.playMusic does this already and accounts for mute

**Review by Aryan-S-Gupta — APPROVED** [review]
Looks Good

**Review by CameronBadman — APPROVED** [review]
LGTM

**Pull request merged**